boot/zephyr/main.c

```diff
@@ -125,8 +129,14 @@ static void do_boot(struct boot_rsp *rsp)
 #ifdef CONFIG_USB
     /* Disable the USB to prevent it from firing interrupts */
     usb_disable();
+#endif
+#if CONFIG_MCUBOOT_CLEANUP_ARM_CORE
+    cleanup_arm_nvic(); /* cleanup NVIC registers */
 #endif
     __set_MSP(vt->msp);
```

Comment:
__set_MSP is not part of the change in this patch, however, I have a question here: **does do_boot() function uses MSP or PSP?** If it uses MSP there might be an issue here. `vt `is a local variable saved in stack. After __set_MSP is changing the stack pointer, normally you  should not rely on the compiler properly de-referencing `vt`, it usually works but some strange compiler settings might make it not work.

I would consider it a dangerous thing, @d3zd3z @nvlsianpu .

Reply:
I think the only way to properly fix that would be to write the set-msp and the jump to reset in assembly. It is concerning to me that another patch has added a __set_CONTROL(0x00) call in between the MSP set and the reset, which might cause 'vt' to get written to the wrong place as well. Given that this is merged, I think this should get its own issue, though.

Reply:
Thinking about this a little more, I don't think this will ever be a problem. Currently the _set calls are inline assembly. If they were ever changed to cause registers to be spilled to the stack, a lot more code than MCUboot would break.

Reply:
@d3zd3z , _set_MSP is inline assembly, that's correct. But vt in line 140 is accessed from C code, using the stack pointer, that is changed via _set_MSP.

The implementation of _set_MSP is as follows:
```
__STATIC_FORCEINLINE void __set_MSP(uint32_t topOfMainStack)
{
  __ASM volatile ("MSR msp, %0" : : "r" (topOfMainStack) : );
}
```
As you can see there is no compiler barrier, either, so stack-pointer based accesses later in the function do not take into account the stack pointer change.